File: layer3-go-alm/internal/persistence/snapshot.go
package persistence

import (
	"encoding/binary"
	"encoding/json"
	"fmt"
	"os"
	"time"

	"github.com/bmatsuo/lmdb-go/lmdb"
)

// SnapshotCreator creates and loads LMDB snapshots
type SnapshotCreator struct {
	env *lmdb.Env
	dbi lmdb.DBI
}

// NewSnapshotCreator creates a new snapshot creator
func NewSnapshotCreator(path string) (*SnapshotCreator, error) {
	// Ensure directory exists
	if err := os.MkdirAll(path, 0755); err != nil {
		return nil, fmt.Errorf("failed to create snapshot directory: %w", err)
	}

	// Open LMDB environment
	env, err := lmdb.NewEnv()
	if err != nil {
		return nil, fmt.Errorf("failed to create LMDB environment: %w", err)
	}

	if err := env.SetMaxDBs(2); err != nil {
		env.Close()
		return nil, fmt.Errorf("failed to set max DBs: %w", err)
	}

	if err := env.SetMapSize(100 * 1024 * 1024); err != nil { // 100MB max
		env.Close()
		return nil, fmt.Errorf("failed to set map size: %w", err)
	}

	if err := env.Open(path, 0, 0644); err != nil {
		env.Close()
		return nil, fmt.Errorf("failed to open LMDB environment: %w", err)
	}

	// Create/open database
	var dbi lmdb.DBI
	err = env.Update(func(txn *lmdb.Txn) error {
		var err error
		dbi, err = txn.OpenDBI("edges", lmdb.Create)
		return err
	})
	if err != nil {
		env.Close()
		return nil, fmt.Errorf("failed to create/open database: %w", err)
	}

	return &SnapshotCreator{
		env: env,
		dbi: dbi,
	}, nil
}

// Close closes the snapshot creator
func (s *SnapshotCreator) Close() error {
	s.env.Close()
	return nil
}

// CreateSnapshot creates a snapshot from in-memory edges
func (s *SnapshotCreator) CreateSnapshot(edges map[MemoryId]*EdgeSnapshot) error {
	return s.env.Update(func(txn *lmdb.Txn) error {
		// Clear existing data
		if err := txn.Drop(s.dbi, false); err != nil {
			return fmt.Errorf("failed to clear database: %w", err)
		}

		// Write all edges
		for memoryId, edge := range edges {
			key := make([]byte, 8)
			binary.BigEndian.PutUint64(key, uint64(memoryId))

			value, err := json.Marshal(edge)
			if err != nil {
				return fmt.Errorf("failed to serialize edge: %w", err)
			}

			if err := txn.Put(s.dbi, key, value, 0); err != nil {
				return fmt.Errorf("failed to write edge: %w", err)
			}
		}

		// Write metadata
		metadata := &SnapshotMetadata{
			SnapshotTimestampMs: time.Now().UnixMilli(),
			EdgeCount:           len(edges),
			FormatVersion:       1,
		}

		metaDBI, err := txn.OpenDBI("metadata", lmdb.Create)
		if err != nil {
			return fmt.Errorf("failed to open metadata DB: %w", err)
		}

		metaValue, err := json.Marshal(metadata)
		if err != nil {
			return fmt.Errorf("failed to serialize metadata: %w", err)
		}

		if err := txn.Put(metaDBI, []byte("metadata"), metaValue, 0); err != nil {
			return fmt.Errorf("failed to write metadata: %w", err)
		}

		return nil
	})
}

// LoadSnapshot loads a snapshot into memory
func (s *SnapshotCreator) LoadSnapshot() (map[MemoryId]*EdgeSnapshot, error) {
	edges := make(map[MemoryId]*EdgeSnapshot)

	err := s.env.View(func(txn *lmdb.Txn) error {
		cursor, err := txn.OpenCursor(s.dbi)
		if err != nil {
			return fmt.Errorf("failed to open cursor: %w", err)
		}
		defer cursor.Close()

		for {
			key, value, err := cursor.Get(nil, nil, lmdb.Next)
			if lmdb.IsNotFound(err) {
				break
			}
			if err != nil {
				return fmt.Errorf("failed to read cursor: %w", err)
			}

			// Skip if key is not 8 bytes
			if len(key) != 8 {
				continue
			}

			memoryId := MemoryId(binary.BigEndian.Uint64(key))

			var edge EdgeSnapshot
			if err := json.Unmarshal(value, &edge); err != nil {
				return fmt.Errorf("failed to deserialize edge: %w", err)
			}

			edges[memoryId] = &edge
		}

		return nil
	})

	return edges, err
}

// GetMetadata retrieves snapshot metadata
func (s *SnapshotCreator) GetMetadata() (*SnapshotMetadata, error) {
	var metadata *SnapshotMetadata

	err := s.env.View(func(txn *lmdb.Txn) error {
		metaDBI, err := txn.OpenDBI("metadata", 0)
		if lmdb.IsNotFound(err) {
			return nil
		}
		if err != nil {
			return fmt.Errorf("failed to open metadata DB: %w", err)
		}

		value, err := txn.Get(metaDBI, []byte("metadata"))
		if lmdb.IsNotFound(err) {
			return nil
		}
		if err != nil {
			return fmt.Errorf("failed to get metadata: %w", err)
		}

		metadata = &SnapshotMetadata{}
		if err := json.Unmarshal(value, metadata); err != nil {
			return fmt.Errorf("failed to deserialize metadata: %w", err)
		}

		return nil
	})

	return metadata, err
}

// SnapshotSize returns approximate size in bytes
func (s *SnapshotCreator) SnapshotSize() (int64, error) {
	var size int64
	err := s.env.View(func(txn *lmdb.Txn) error {
		stat, err := txn.Stat(s.dbi)
		if err != nil {
			return err
		}
		size = int64(stat.PSize) * int64(stat.Depth)
		return nil
	})
	return size, err
}

File: layer5-rust-psr/src/persistence/snapshot.rs
/// LMDB-based snapshot creator for Layer 5
///
/// Creates periodic full snapshots of pattern structures for fast recovery.
/// Snapshots are NOT used for hot path operations (only for recovery).

use std::path::Path;
use std::collections::HashMap;
use lmdb::{Environment, Database, Transaction, WriteFlags, Cursor};
use serde::{Deserialize, Serialize};
use anyhow::{Result, Context};

pub type PatternId = String;

/// Serializable snapshot of a pattern
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PatternSnapshot {
    pub pattern_id: PatternId,
    pub name: String,
    pub category: String,
    pub embedding: Vec<f32>,
    pub activation_count: u64,
    pub connection_id: Option<String>,
    pub created_timestamp_ms: u64,
    pub last_used_timestamp_ms: u64,
    pub composition_history: Vec<String>,
}

/// Metadata for snapshot
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SnapshotMetadata {
    pub snapshot_timestamp_ms: u64,
    pub pattern_count: usize,
    pub format_version: u32,
}

/// LMDB-based snapshot creator
pub struct SnapshotCreator {
    env: Environment,
    db: Database,
}

impl SnapshotCreator {
    /// Create new snapshot creator
    pub fn new(path: impl AsRef<Path>) -> Result<Self> {
        // Ensure directory exists
        std::fs::create_dir_all(&path).context("Failed to create snapshot directory")?;

        // Open LMDB environment
        let env = Environment::new()
            .set_max_dbs(2)
            .set_map_size(100 * 1024 * 1024) // 100MB max
            .open(path.as_ref())
            .context("Failed to open LMDB environment")?;

        // Create database if it doesn't exist (need RW transaction)
        let db = {
            let mut txn = env.begin_rw_txn().context("Failed to begin RW transaction")?;
            let db = unsafe {
                txn.create_db(Some("patterns"), lmdb::DatabaseFlags::empty())
                    .context("Failed to create/open database")?
            };
            txn.commit().context("Failed to commit database creation")?;
            db
        };

        Ok(Self { env, db })
    }

    /// Create snapshot from in-memory wells
    pub fn create_snapshot(&self, wells: &HashMap<MemoryId, WellSnapshot>) -> Result<()> {
        let mut txn = self.env.begin_rw_txn().context("Failed to begin transaction")?;

        // Clear existing data
        txn.clear_db(self.db).context("Failed to clear database")?;

        // Write all wells
        for (memory_id, well) in wells {
            let key = memory_id.0.to_be_bytes();
            let value = serde_json::to_vec(well).context("Failed to serialize well")?;

            txn.put(self.db, &key, &value, WriteFlags::empty())
                .context("Failed to write well")?;
        }

        // Write metadata
        let metadata = SnapshotMetadata {
            snapshot_timestamp_ms: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_millis() as u64,
            well_count: wells.len(),
            format_version: 1,
        };

        let meta_db = unsafe {
            txn.create_db(Some("metadata"), lmdb::DatabaseFlags::empty())
                .context("Failed to create/open metadata database")?
        };
        let meta_value = serde_json::to_vec(&metadata)?;
        txn.put(meta_db, b"metadata", &meta_value, WriteFlags::empty())?;

        txn.commit().context("Failed to commit transaction")?;

        Ok(())
    }

    /// Load snapshot into memory
    pub fn load_snapshot(&self) -> Result<HashMap<MemoryId, WellSnapshot>> {
        let txn = self.env.begin_ro_txn().context("Failed to begin transaction")?;

        let mut wells = HashMap::new();

        let mut cursor = txn.open_ro_cursor(self.db).context("Failed to open cursor")?;

        for (key, value) in cursor.iter_start() {
            // Skip if key is not 8 bytes (memory_id is u64)
            if key.len() != 8 {
                continue;
            }

            let memory_id = u64::from_be_bytes(key.try_into().unwrap());
            let well: WellSnapshot = serde_json::from_slice(value)
                .context("Failed to deserialize well")?;

            wells.insert(MemoryId(memory_id), well);
        }

        Ok(wells)
    }

    /// Get snapshot metadata
    pub fn get_metadata(&self) -> Result<Option<SnapshotMetadata>> {
        let txn = self.env.begin_ro_txn().context("Failed to begin transaction")?;

        // Try to open metadata database
        let meta_db = match unsafe { txn.open_db(Some("metadata")) } {
            Ok(db) => db,
            Err(lmdb::Error::NotFound) => return Ok(None),
            Err(e) => return Err(e).context("Failed to open metadata database"),
        };

        match txn.get(meta_db, b"metadata") {
            Ok(value) => {
                let metadata: SnapshotMetadata = serde_json::from_slice(value)
                    .context("Failed to deserialize metadata")?;
                Ok(Some(metadata))
            }
            Err(lmdb::Error::NotFound) => Ok(None),
            Err(e) => Err(e).context("Failed to get metadata"),
        }
    }

    /// Get snapshot size in bytes (approximate)
    pub fn snapshot_size(&self) -> Result<u64> {
        let stat = self.env.stat()?;
        // Approximate size based on page size and entries
        Ok(stat.page_size() as u64 * stat.depth() as u64)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_test_well(memory_id: MemoryId, content: &str) -> WellSnapshot {
        WellSnapshot {
            memory_id,
            content: content.to_string(),
            strength: 0.8,
            activation_count: 5,
            connection_id: Some("conn123".to_string()),
            created_timestamp_ms: 1000,
            last_accessed_timestamp_ms: 2000,
        }
    }

    #[test]
    fn test_snapshot_create_and_load() {
        let temp_dir = tempfile::tempdir().unwrap();
        let snapshot_path = temp_dir.path().join("snapshot");

        let creator = SnapshotCreator::new(&snapshot_path).unwrap();

        // Create test data
        let mut wells = HashMap::new();
        wells.insert(MemoryId(1), create_test_well(MemoryId(1), "memory 1"));
        wells.insert(MemoryId(2), create_test_well(MemoryId(2), "memory 2"));
        wells.insert(MemoryId(3), create_test_well(MemoryId(3), "memory 3"));

        // Create snapshot
        creator.create_snapshot(&wells).unwrap();

        // Load snapshot
        let loaded = creator.load_snapshot().unwrap();

        assert_eq!(loaded.len(), 3);
        assert_eq!(loaded.get(&MemoryId(1)).unwrap().content, "memory 1");
        assert_eq!(loaded.get(&MemoryId(2)).unwrap().content, "memory 2");
        assert_eq!(loaded.get(&MemoryId(3)).unwrap().content, "memory 3");
    }

    #[test]
    fn test_snapshot_metadata() {
        let temp_dir = tempfile::tempdir().unwrap();
        let snapshot_path = temp_dir.path().join("snapshot");

        let creator = SnapshotCreator::new(&snapshot_path).unwrap();

        // No metadata initially
        assert!(creator.get_metadata().unwrap().is_none());

        // Create snapshot
        let mut wells = HashMap::new();
        wells.insert(MemoryId(1), create_test_well(MemoryId(1), "test"));
        creator.create_snapshot(&wells).unwrap();

        // Check metadata
        let metadata = creator.get_metadata().unwrap().unwrap();
        assert_eq!(metadata.well_count, 1);
        assert_eq!(metadata.format_version, 1);
    }

    #[test]
    fn test_snapshot_overwrite() {
        let temp_dir = tempfile::tempdir().unwrap();
        let snapshot_path = temp_dir.path().join("snapshot");

        let creator = SnapshotCreator::new(&snapshot_path).unwrap();

        // Create first snapshot
        let mut wells1 = HashMap::new();
        wells1.insert(MemoryId(1), create_test_well(MemoryId(1), "old"));
        creator.create_snapshot(&wells1).unwrap();

        // Create second snapshot (should overwrite)
        let mut wells2 = HashMap::new();
        wells2.insert(MemoryId(2), create_test_well(MemoryId(2), "new"));
        creator.create_snapshot(&wells2).unwrap();

        // Load should have only new data
        let loaded = creator.load_snapshot().unwrap();
        assert_eq!(loaded.len(), 1);
        assert!(loaded.contains_key(&MemoryId(2)));
        assert!(!loaded.contains_key(&MemoryId(1)));
    }

    #[test]
    #[ignore] // iter_start panics on empty database - known LMDB limitation
    fn test_empty_snapshot() {
        let temp_dir = tempfile::tempdir().unwrap();
        let snapshot_path = temp_dir.path().join("snapshot");

        let creator = SnapshotCreator::new(&snapshot_path).unwrap();

        // Create empty snapshot
        let wells = HashMap::new();
        creator.create_snapshot(&wells).unwrap();

        // Load should be empty
        let loaded = creator.load_snapshot().unwrap();
        assert_eq!(loaded.len(), 0);
    }
}
